chain/chain/src/chain.rs

```diff
@@ -2957,29 +3014,41 @@ impl Chain {
         debug!(target:"catchup", "catch up blocks: pending blocks: {:?}, processed {:?}, scheduled: {:?}, done: {:?}",
                blocks_catch_up_state.pending_blocks, blocks_catch_up_state.processed_blocks.keys().collect::<Vec<_>>(),
                blocks_catch_up_state.scheduled_blocks, blocks_catch_up_state.done_blocks.len());
+        let mut processed_blocks = HashMap::new();
         for (queued_block, results) in blocks_catch_up_state.processed_blocks.drain() {
-            match self.block_catch_up_postprocess(&queued_block, results) {
-                Ok(_) => {
-                    let mut saw_one = false;
-                    for next_block_hash in self.store.get_blocks_to_catchup(&queued_block)?.clone()
-                    {
-                        saw_one = true;
-                        blocks_catch_up_state.pending_blocks.push(next_block_hash);
+            // If this block is parent of some blocks in processing that need to be caught up,
+            // we can't mark this block as done yet because these blocks haven't been added to
+            // the store as blocks to be caught up yet. If we mark this block as done right now,
+            // these blocks will never get caught up. So we add these blocks back to the processed_blocks
+            // queue.
+            if self.blocks_in_processing.has_blocks_to_catch_up(&queued_block) {
+                processed_blocks.insert(queued_block, results);
```

Comment:
Heh, I must confess, while I nodded along while reading the comment, I still don't understand what this actually talks about. Not actionable for this PR, but I wonder if there's some diagram somewhere depicting the state transitions of block's CrytpoHash? 

That is, I think I understand that, once we've learnd of a block hash, it hash to go through various stages until we incorporate it in chain proper. There's a lot of fuzzy-sounding terminology surrounding this (block is catching up/caught up/pre-processsed/processed/applied, etc), but, at least in my head, it doesn't yet add up to a clear state-transition diagram. 

(the second step after the diagram would be to re-ify this in code: I'd really love to see one file per "step": `pre_processing.rs`, `processing.rs`, `catchup.rs`. It would also be interesting to tag crytpo-hashes at the type level: `struct PreProcessed(CryptoHash)`, `struct Processed(CryptoHash)`, `fn process(PreProcessed) -. Processed`). 

Sorry for the vague rambling :)

Reply:
Ah, I guess 

![image](https://user-images.githubusercontent.com/1711539/176435708-c7a2100f-6892-4360-874f-673f56f313c0.png)

is roughly what I am looking for, but it seems like that is missing some steps

Reply:
Made some progress in furthering my own understanding: https://nearinc.atlassian.net/wiki/spaces/EAP/pages/89718785/How+the+block+gets+made

Reply:
> Ah, I guess
> 
> ![image](https://user-images.githubusercontent.com/1711539/176435708-c7a2100f-6892-4360-874f-673f56f313c0.png)
> 
> is roughly what I am looking for, but it seems like that is missing some steps

Yes and the catching up logic is not included here. There is another doc for that though https://nearinc.atlassian.net/wiki/spaces/EAP/pages/32375041/How+Sync+works